### PR TITLE
[MRG] Encoding problems

### PR DESCRIPTION
**Pull request by Titan-C**
The filesystem's encoding has generated quite some problems  #257, #259, #260, #269
I started this branch long ago, I'll bring it back to live now. This starts the test on Travis-CI for the ASCII encoding.
It does not work yet.

### CODE REVIEW

**Comment by lesteve** [comment]
It'd be nice to fix this indeed. Your last commit status is green, does that mean that this is ready for review?

**Comment by Titan-C** [comment]
> It'd be nice to fix this indeed. Your last commit status is green, does that mean that this is ready for review?

Not really. I just arrived to an easy solution. Enforcing reading all
files as byte strings. But I actually want to refactor and put some more
test. This solution is easy and wont bite us back as long as no one
declares function names with characters outside ASCII before terminate
support for Python2.7. But I'm still a bit worried about comments as
normal strings.

I also need a new test that verifies code blocks are correctly parsed
for embedding documentation links, to avoid #259 happening again.

**Comment by lesteve** [comment]
> I also need a new test that verifies code blocks are correctly parsed for embedding documentation links, to avoid #259 happening again.

I think there is a test for that already:
https://github.com/sphinx-gallery/sphinx-gallery/blob/587198a79c6e602e7a0be97f1f06c9ab57a123c2/sphinx_gallery/tests/test_full.py#L48-L64

Although for some reason the matplotlib check was not passing on Travis which is why it is commented out. I could not reproduce, see https://github.com/sphinx-gallery/sphinx-gallery/pull/308#discussion_r146858672 for more details.

**Comment by larsoner** [comment]
> Although for some reason the matplotlib check was not passing on Travis 

I suspect this was due to old `sphinx` being used

**Comment by Titan-C** [comment]
This is finally working. I even found some extra problems that could be originated by different encodings, characters in use and the source file config. I have added comments explaining such events. I also have generated a new example using unicode emoji just to stress test this setup.
https://184-26191147-gh.circle-artifacts.com/0/home/ubuntu/sphinx-gallery/alabaster_html/auto_examples/plot_unicode_everywhere.html

**Comment by Titan-C** [comment]
Are there any comments on this? Can I merge?
@pllim @NelleV does this solve your issues #253 & #269 ?

**Comment by pllim** [comment]
@Titan-C , I don't have time to test this right now but I can give it a try in the future when a new release with this fix is in. Thanks!

File: sphinx_gallery/py_source_parser.py [hunk 1]
```diff
@@ -19,30 +19,57 @@
...
         content = fid.read()
     # change from Windows format to UNIX for uniformity
     content = content.replace(b'\r\n', b'\n')
 
     try:
         node = ast.parse(content)
+        return node, content.decode('utf-8')
     except SyntaxError:
-        return SYNTAX_ERROR_DOCSTRING, content.decode('utf-8'), 1
+        return SyntaxError, content.decode('utf-8')
```
Comment (lesteve) [inline sphinx_gallery/py_source_parser.py]: This feels a bit kludgy to return an exception class rather than a node. Why don't you SyntaxError through and catch it in `get_docstring_and_rest` and have an except clause at the `parse_file_source` call?

File: sphinx_gallery/tests/test_backreferences.py [hunk 2]
```diff
@@ -89,4 +78,30 @@ def test_identify_names():
...
+from a.b import c
+import d as e
+print(c)
+e.HelloWorld().f.g
+"""
+    expected = {'c': {'name': 'c', 'module': 'a.b', 'module_short': 'a.b'},
+                'e.HelloWorld': {'name': 'HelloWorld', 'module': 'd', 'module_short': 'd'}}
+
+    import locale
+    print(locale.getlocale())
```
Comment (lesteve) [inline sphinx_gallery/tests/test_backreferences.py]: Looks like debugging print statements that you forgot to remove.

File: sphinx_gallery/tests/test_backreferences.py [hunk 3]
```diff
@@ -89,4 +78,30 @@ def test_identify_names():
...
+
+    with tempfile.NamedTemporaryFile('wb', delete=False) as f:
```
Comment (lesteve) [inline sphinx_gallery/tests/test_backreferences.py]: Using pytest `tmpdir` fixture is a lot more neat for this kind of things.

File: sphinx_gallery/py_source_parser.py [hunk 4]
```diff
@@ -19,30 +19,57 @@
 """
 
 
-def get_docstring_and_rest(filename):
-    """Separate `filename` content between docstring and the rest
+def parse_file_source(filename):
```
Comment (lesteve) [inline sphinx_gallery/py_source_parser.py]: Maybe `parse_python_file` or `parse_source_file`?

File: sphinx_gallery/py_source_parser.py [hunk 5]
```diff
@@ -19,30 +19,57 @@
...
     try:
         node = ast.parse(content)
+        return node, content.decode('utf-8')
     except SyntaxError:
-        return SYNTAX_ERROR_DOCSTRING, content.decode('utf-8'), 1
+        return SyntaxError, content.decode('utf-8')
+
+
+def get_docstring_and_rest(filename):
+    """Separate `filename` content between docstring and the rest
```
Comment (lesteve) [inline sphinx_gallery/py_source_parser.py]: I think you need to use double backticks if you want this to be fixed-width in the HTML reference documentation. Same thing for a few single backticks that you use in a few different places.

File: sphinx_gallery/tests/test_py_source_parser.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+# License: 3-clause BSD
+
+from __future__ import division, absolute_import, print_function
+import sphinx_gallery.py_source_parser as sg
+
+
+def test_get_docstring_and_rest():
+
+    docstring, rest, lineno = sg.get_docstring_and_rest(
+        'sphinx_gallery/tests/unicode.sample')
```
Comment (lesteve) [inline sphinx_gallery/tests/test_py_source_parser.py]: Can you write the content into a temporary file instead using `tmpdir`? This way the test is a bit easier to understand.

**Comment by lesteve** [comment]
@Titan-C I am looking at this, please don't pull the trigger yet ;-) !

File: sphinx_gallery/py_source_parser.py (hunk 1 above)
Comment (Titan-C) [reply sphinx_gallery/py_source_parser.py]: I agree is not the nicest way. I ended up using this, because I need this function in few more places and I rely those places always having access to the decoded content and to know if the parsing was successful or not.

Comment (lesteve) [reply sphinx_gallery/py_source_parser.py]: I saw that but I would rather let the SyntaxError go through and catch it where you call `get_docstring_and_rest`. Arguably you could have a SyntaxError somewhere else in your code but I don't think this is that bad. If we care about that we could catch the SyntaxError in `parse_source_file` and raise a custom exception e.g. ParsingError (or a better name) then catch ParsingError where parse_source_file` is called.

Comment (Titan-C) [reply sphinx_gallery/py_source_parser.py]: This is on the boundaries of my knowledge. But if I raise the exception how do I get access to the returned value of content?

Comment (lesteve) [reply sphinx_gallery/py_source_parser.py]: Just in case I am missing and the exception approach does not work, can you not return None instead?

File: sphinx_gallery/tests/test_py_source_parser.py (hunk 6 above)
Comment (Titan-C) [reply sphinx_gallery/tests/test_py_source_parser.py]: Not really. If you check on the previous file there is `test_identify_names2` there I indeed write to file. And for everything to work correctly in py2 & py3 I write it as bytes. But in this case and the function `test_identify_names` I really want to read a file from disk, the file is identical. But is the 2 places I need to be sure the file read from disk is correctly parsed.

File: sphinx_gallery/tests/test_backreferences.py (hunk 3 above)
Comment (Titan-C) [reply sphinx_gallery/tests/test_backreferences.py]: I just did this of using the fixture, but all tests jobs in travis (except 1) failed

File: sphinx_gallery/tests/test_py_source_parser.py (hunk 6 above)
Comment (lesteve) [reply sphinx_gallery/tests/test_py_source_parser.py]: It seems like this does work:

```py

content = b'\n'.join(
    [b'',
     b'# -*- coding: utf-8 -*-',
     b"'''",
     b'\xc3\x9anicode in header',
     b'=================',
     b'',
     b'U\xc3\xb1icode in description',
     b"'''",
     b'',
     b'# Code source: \xc3\x93scar N\xc3\xa1jera',
     b'# License: BSD 3 clause',
     b'',
     b'import os',
     b"path = os.path.join('a','b')",
     b'',
     b"a = 'hei\xc3\x9f'  # Unicode string",
     b'',
     b'import sphinx_gallery.back_references as br',
     b'br.identify_names',
     b'',
     b'from sphinx_gallery.back_references import identify_names',
     b'identify_names ',
     b''])

open('/tmp/unicode.sample', 'wb').write(content)

# use tmpdir to get a file here instead
open('/tmp/unicode.sample', 'wb').write(content)
```

File: sphinx_gallery/py_source_parser.py (hunk 1 above)
Comment (Titan-C) [reply sphinx_gallery/py_source_parser.py]: Of course. I can return None

File: sphinx_gallery/tests/test_backreferences.py (hunk 3 above)
Comment (Titan-C) [reply sphinx_gallery/tests/test_backreferences.py]: Any info on this. It's recent commit https://github.com/sphinx-gallery/sphinx-gallery/pull/311/commits/f2f2d04e4fd7401b1ec1ffe6833716505c65c166

File: sphinx_gallery/tests/test_py_source_parser.py (hunk 6 above)
Comment (Titan-C) [reply sphinx_gallery/tests/test_py_source_parser.py]: Are you sure this is simpler than having a unicode heavy python file on disk? Also, I use this file twice, one to test that despite the unicode I `get_docstring_and_rest` and one to test that despite the unicode I `identify_names` of the used functions.

Comment (lesteve) [reply sphinx_gallery/tests/test_py_source_parser.py]: > Are you sure this is simpler than having a unicode heavy python file on disk?

Yes, this is a lot more clear because you see exactly what you are writing to the file. Otherwise how do we know how you generated your file. Also test files like this don't get included in the installed package which mean that you can not run the tests on an installed package to make sure that they run. For example this is a problem we already have for sphinx_gallery/tests/reference_parse.txt.

If you use the file content in two tests you can put that in a helper function.

Comment (Titan-C) [reply sphinx_gallery/tests/test_py_source_parser.py]: OK then. I implemented a pytest fixture to generate this sample file

**Comment by Titan-C** [comment]
This is again ready for review. 
Travis on ubuntu fails because the system version of six is too old and pytest coverage refuses to load. But this is a particular situation because
`python setup.py test` passes as it installs *for the test* a resent version of six
`pytest sphinx_gallery` fails because it uses the old system version.
Is everyone fine if I pip install a more resent version of six in this test environment? Other ideas?

**Comment by lesteve** [comment]
I saw that a few days ago (basically we have a failure on master at the moment because of this). I just opened #325 to fix this.

I guess (I did not check in details) that this is one of the reason that @choldgraf had troubles in #244. Installing sphinx 1.5 probably pulls in a more recent version of six ...

**Comment by lesteve** [comment]
I added an entry in CHANGES.rst, @Titan-C you probably understand the problem more than me, so feel free to improve/fix it!

Other than this, I think this is ready to be merged, great work on a nasty long-standing issue!

**Comment by lesteve** [comment]
Merging this one, thanks a lot @Titan-C!

**Comment by ImportanceOfBeingErnest** [comment]
When will a version containing this fix be released?    
(Building the matplotlib docs on windows currently blocks on this.)

**Comment by choldgraf** [comment]
Relatively soon, I believe. We are spot-checking the remaining issues before a release. @lesteve just opened a PR for one and I'll plan to hit the others later this week or next. As always, PRs are welcome ;-)

**Comment by jni** [comment]
I just want to note that this PR also fixed an unclosed file warning when building docs:

```pytb
/home/travis/venv/lib/python3.4/site-packages/sphinx_gallery/backreferences.py:144:
ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/travis/build/scikit-image/scikit-image/doc/source/auto_examples/color_exposure/plot_ihc_color_separation.py' mode='r' encoding='UTF-8'>
```

See e.g. https://travis-ci.org/scikit-image/scikit-image/jobs/372237529#L6134

I searched issues and PRs for ResourceWarning and couldn't find anything, so I presumed it was unfixed and *almost* raised a new issue. =)

Thanks for the great work!

**Comment by jni** [comment]
[This](https://github.com/sphinx-gallery/sphinx-gallery/pull/311/files#diff-33215ef937b7c3d86fe0a23425c585e4L144) was the offending line, now properly context-managed elsewhere:

```python
example_code_obj = identify_names(open(example_file).read())
```
